File: server/gamestates/WaitingForPromotionState.cpp
#include "WaitingForPromotionState.h"
#include "DropState.h"
#include "../Game.h"
#include "../Player.h"
#include "../exceptions/NoSuchGameException.h"
#include "../exceptions/InvalidActionException.h"
#include "../ChessMatrix.h"

void WaitingForPromotionState::promote(PawnPromotionStrategy* strategy)
{
    Piece* newPiece = strategy->createPiece();
    
    newPiece->set(_pawn->getId(),_pawn->isWhite(),_pawn->getPos(),
        _game,_pawn->getForward());

    _game->setState(_playingState);
    _game->tickTurn();

    _game->getMatrix()->set(newPiece->getPos() ,newPiece);

    delete this;
}

bool WaitingForPromotionState::drop (const std::string& userId)
{
    bool white;
    
    if (_game->playerW()->user() == userId)
        white = true;
    
    else if (_game->playerB()->user() == userId)
        white = false;
    
    else 
        throw NoSuchGameException ();
        
    _game->setState(new DropState (_game, white));
    
    delete _playingState;
    delete this;
    
    return false;
}

File: client/js/gamestates/WaitingForPromotionState.js
class WaitingForPromotionState extends GameState {
    constructor(playingState, game, pawn) {
        this.game = game;
        this.playingState = playingState;
        this.pawn = pawn;
    }

    accept(visitor) {
        visitor.visitWaiting(this);
    }
    move(origin, dest) {
        throw new InvalidActionException();
    }
    promote(piece) {
        piece.set(this.pawn.id,this.pawn.white,this.pawn.pos,
            this.game, this.pawn.forward);
        
        this.game.setState(this.playingState);
        this.game.tickTurn();
        this.game.chessMatrix.set(piece.pos,piece);
    }
    
    checkMove(origin, dest) {
        throw new InvalidActionException();
    }
}